### PR TITLE
Add flag for size based retention

### PR DESCRIPTION
**Pull request by gouthamve**
@krasi-georgiev Could you commit the vendor update, it's taking ages upon ages for me

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
decided to update tsdb in a separate PR to keep the git history separate and reverted the tsdb update commits

https://github.com/prometheus/prometheus/pull/5110

File: storage/tsdb/tsdb.go [hunk 1]
```diff
@@ -195,6 +208,27 @@ func Open(path string, l log.Logger, r prometheus.Registerer, opts *Options) (*t
 	return db, nil
 }
 
+// ChooseRetention is some roundabout code to support both RetentionDuration and Retention (for different flags).
+// If Retention is 15d, then it means that the default value is set and the value of RetentionDuration is used.
+func ChooseRetention(oldFlagDuration, newFlagDuration model.Duration) model.Duration {
+	defaultDuration, err := model.ParseDuration("15d")
```
Comment (codesome) [inline storage/tsdb/tsdb.go]: I see `"15d"` being used in 4 places now (1 more in test). Should be have a single variable for this?

**Comment by brancz** [comment]
--storage.tsdb.retention can be deprecated but must continue to work. The new flag can have precedence, but the old one must continue to work until 3.0.

**Comment by brancz** [comment]
Agreed with Simon.

**Comment by gouthamve** [comment]
Yup, @brancz that is what is happening. And yup, I moved the logic into main.go. It's better, thanks! @simonpasquier

File: storage/tsdb/tsdb.go [hunk 2]
```diff
@@ -33,6 +33,8 @@ import (
 // ErrNotReady is returned if the underlying storage is not ready yet.
 var ErrNotReady = errors.New("TSDB not ready")
 
+const DefaultRetentionDuration = "15d"
```
Comment (simonpasquier) [inline storage/tsdb/tsdb.go]: not needed anymore

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -244,15 +257,25 @@ func main() {
...
-		cfg.tsdb.MaxBlockDuration = cfg.tsdb.Retention / 10
+		cfg.tsdb.MaxBlockDuration = cfg.tsdb.RetentionDuration / 10
 	}
 
 	promql.LookbackDelta = time.Duration(cfg.lookbackDelta)
 	promql.SetDefaultEvaluationInterval(time.Duration(config.DefaultGlobalConfig.EvaluationInterval))
 
 	logger := promlog.New(&cfg.promlogConfig)
 
+	defaultDuration, err := model.ParseDuration(tsdb.DefaultRetentionDuration)
```
Comment (simonpasquier) [inline cmd/prometheus/main.go]: As we do the same in `chooseRetention()`, I'd use a global `mode.Duration` variable and move this to the `init()` function.

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -757,3 +780,24 @@ func sendAlerts(s sender, externalURL string) rules.NotifyFunc {
...
+		panic(err)
+	}
+
+	retention := oldFlagDuration
+	if retention == defaultDuration {
+		retention = newFlagDuration
+	}
+
+	// Further if both the flags are set, then RetentionDuration takes precedence.
+	if newFlagDuration != defaultDuration {
```
Comment (brancz) [inline cmd/prometheus/main.go]: I'm not sure this yields the expected behavior, when we say the new flag takes precedence. If the new flag explicitly specifies 15 days and the old flag specifies 30 days, then the old one will be chosen.

Comment (gouthamve) [reply cmd/prometheus/main.go]: Hmm, fair. Not sure how to deal with the case though :/ Detecting if the flag has been set or if it's just the default value is next to impossible.

Comment (brancz) [reply cmd/prometheus/main.go]: I think just setting the flag description to explain this rectifies the situation.

"If this flag is set to anything other than the default, it takes precedence."

Comment (gouthamve) [reply cmd/prometheus/main.go]: Done

**Comment by brancz** [comment]
lgtm :+1:

**Comment by gouthamve** [comment]
@simonpasquier @brancz Feedback addressed, PTAL.

Thanks for all the rounds of reviews, sorry for the trouble!

**Pull request merged**